embassy-stm32/src/pka/mod.rs

```diff
@@ -0,0 +1,830 @@
+        // Calculate the maximum size of operands based on leading zeros
+        let operand_size = 8 * (a.len() as u32);
+
+        self.write_to_ram_u32(ModularMultiplication::OperandLength.value(), &operand_size);
+
+        self.write_to_ram(ModularMultiplication::Modulus.value(), modulus);
+        self.write_to_ram(ModularMultiplication::OperandA.value(), &a);
+        self.write_to_ram(ModularMultiplication::OperandB.value(), &b);
+
+        let mode = PkaOperationMode::MontgomeryMultiplication as u8;
+        self.set_mode(mode);
+
+        self.start().await;
+
+        self.read_from_ram(ModularMultiplication::Result.value(), result);
+    }
+
+    /// Calculate the modular inversion of a number
+    pub async fn mod_inverse(&mut self, a: &[u8], modulus: &[u8], result: &mut [u8]) {
+        // Calculate the operand size.
+        let operand_size = 8 * (modulus.len() as u32);
+
+        self.write_to_ram_u32(ModularInversion::OperandLength.value(), &operand_size);
+
+        self.write_to_ram(ModularInversion::OperandA.value(), &a);
+        // This is not an error. Once in a while modulus is written to different address
+        self.write_to_ram(ModularInversion::Modulus.value(), modulus);
+
+        let mode = PkaOperationMode::ModularInversion as u8;
+        self.set_mode(mode);
+
+        self.start().await;
+
+        self.read_from_ram(ModularInversion::Result.value(), result);
+    }
+
+    /// Calculate the modular exponentiation
+    pub async fn mod_exp(
+        &mut self,
+        operand: &[u8],
+        exponent: &[u8],
+        modulus: &[u8],
+        r2_mod_n: &[u8],
+        result: &mut [u8],
+    ) {
+        let exponent_size = 8 * exponent.len() as u32 - exponent[0].leading_zeros() as u32;
+
+        // Calculate the operand size
+        let operand_size = 8 * modulus.len() as u32;
+
+        self.write_to_ram_u32(ModularExponentiation::ExponentLength.value(), &exponent_size);
+        self.write_to_ram_u32(ModularExponentiation::OperandLength.value(), &operand_size);
+
+        self.write_to_ram(ModularExponentiation::OperandA.value(), operand);
+        self.write_to_ram(ModularExponentiation::Exponent.value(), exponent);
+        self.write_to_ram(ModularExponentiation::Modulus.value(), modulus);
+        self.write_to_ram(ModularExponentiation::R2ModN.value(), r2_mod_n);
+
+        let mode = PkaOperationMode::ModularExponentiationOnly as u8;
+        self.set_mode(mode);
+
+        self.start().await;
+
+        self.read_from_ram(ModularExponentiation::Result.value(), result);
+    }
+
+    /// Calculate A * B
+    pub async fn arithmetic_multiply(&mut self, a: &[u8], b: &[u8], result: &mut [u8]) {
+        // Calculate the operand size
+
+        let operand_size = 8 * (a.len() as u32);
+
+        self.write_to_ram_u32(ArithmeticMultiplication::OperandLength.value(), &operand_size);
+        self.write_to_ram(ArithmeticMultiplication::OperandA.value(), &a);
```

Comment:
You can replace `&a` with just `a` here. 

There are more issues like that. Please run `cargo clippy -- -D warnings` in `embassy-stm32`.